File: hybrid-agent/routes/medical_consultation.py
import logging
from datetime import datetime
from typing import Optional, Dict, Any, List
from fastapi import APIRouter, HTTPException, Depends, Request
from pydantic import BaseModel, Field
import asyncio

from core.auth_service import get_current_user
from core.exceptions import ValidationError, ProcessingError
from routes.chat import chat_with_documents
from routes.embed import generate_embedding

logger = logging.getLogger("txagent")

router = APIRouter()

# Request/Response models
class MedicalConsultationRequest(BaseModel):
    query: str = Field(..., description="Medical question or concern")
    context: Optional[Dict[str, Any]] = Field(default=None, description="Additional context including user_profile and conversation_history")
    session_id: Optional[str] = Field(default=None, description="Session identifier")
    preferred_agent: Optional[str] = Field(default="txagent", description="Preferred AI agent for consultation (txagent or openai)")
    
class MedicalConsultationResponse(BaseModel):
    response: Dict[str, Any]
    safety: Dict[str, Any]
    recommendations: Dict[str, Any]
    processing_time_ms: int
    session_id: str
    agent_id: str

# Emergency keywords for detection
EMERGENCY_KEYWORDS = [
    'chest pain', 'difficulty breathing', 'severe bleeding', 'unconscious',
    'heart attack', 'stroke', 'seizure', 'severe allergic reaction',
    'suicidal thoughts', 'overdose', 'can\'t breathe', 'choking',
    'severe headache', 'loss of consciousness', 'severe abdominal pain',
    'severe burns', 'poisoning', 'drug overdose', 'suicide', 'kill myself'
]

def detect_emergency(text: str) -> Dict[str, Any]:
    """Detect emergency keywords in text"""
    lower_text = text.lower()
    detected_keywords = [keyword for keyword in EMERGENCY_KEYWORDS if keyword in lower_text]
    
    return {
        "is_emergency": len(detected_keywords) > 0,
        "confidence": "high" if detected_keywords else "low",
        "detected_keywords": detected_keywords
    }

@router.post("/medical-consultation", response_model=MedicalConsultationResponse)
async def medical_consultation(
    request: MedicalConsultationRequest,
    current_user: Dict[str, Any] = Depends(get_current_user)
):
    """
    Process medical consultation request with emergency detection and AI response
    Supports both doctor (general medical RAG) and patient (personalized) contexts
    """
    start_time = datetime.now()
    user_id = current_user["user_id"]
    
    logger.info(f"🏥 Medical consultation request from user: {user_id}")
    logger.info(f"🔍 Query preview: {request.query[:100]}...")
    logger.info(f"🤖 Preferred agent: {request.preferred_agent}")
    
    # Determine context type based on presence of user_profile
    has_user_profile = bool(request.context and request.context.get("user_profile"))
    context_type = "patient" if has_user_profile else "doctor"
    
    logger.info(f"👤 Context type: {context_type} (has_user_profile: {has_user_profile})")
    
    try:
        # Validate input
        if not request.query or not request.query.strip():
            raise ValidationError("Query is required and cannot be empty")
        
        # Emergency detection (applies to both doctor and patient contexts)
        emergency_check = detect_emergency(request.query)
        
        if emergency_check["is_emergency"]:
            logger.warning(f"🚨 Emergency detected for user {user_id}: {emergency_check['detected_keywords']}")
            
            processing_time = int((datetime.now() - start_time).total_seconds() * 1000)
            
            return MedicalConsultationResponse(
                response={
                    "text": "I've detected that you may be experiencing a medical emergency. Please contact emergency services immediately (call 911) or go to the nearest emergency room. This system cannot provide emergency medical care.",
                    "confidence_score": 0.95,
                    "sources": []
                },
                safety={
                    "emergency_detected": True,
                    "disclaimer": "This is not a substitute for professional medical advice. In case of emergency, contact emergency services immediately.",
                    "urgent_care_recommended": True
                },
                recommendations={
                    "suggested_action": "Contact emergency services immediately (911)",
                    "follow_up_questions": []
                },
                processing_time_ms=processing_time,
                session_id=request.session_id or f"emergency-{user_id}",
                agent_id="emergency_system"
            )
        
        # Process consultation based on preferred agent and context
        logger.info(f"💬 Processing {context_type} consultation for user: {user_id} with {request.preferred_agent}")
        
        # Use the enhanced chat functionality with agent selection
        chat_response = await chat_with_documents(
            query=request.query,
            history=request.context.get("conversation_history", []) if request.context else [],
            top_k=5,
            temperature=0.7,
            stream=False,
            current_user=current_user,
            preferred_agent=request.preferred_agent,
            context=request.context  # Pass full context including user_profile
        )
        
        processing_time = int((datetime.now() - start_time).total_seconds() * 1000)
        
        # Determine appropriate disclaimer based on context type
        if context_type == "patient":
            disclaimer = "This personalized information is for educational purposes only and is not a substitute for professional medical advice, diagnosis, or treatment. Always consult with your healthcare provider."
            suggested_action = "Consult with your healthcare provider for personalized medical advice"
        else:
            disclaimer = "This information is for educational purposes only and is not a substitute for professional medical advice, diagnosis, or treatment."
            suggested_action = "Use this information to support clinical decision-making in conjunction with professional medical judgment"
        
        logger.info(f"✅ Medical consultation completed for user: {user_id} in {processing_time}ms using {chat_response.get('agent_id', request.preferred_agent)}")
        
        return MedicalConsultationResponse(
            response={
                "text": chat_response["response"],
                "sources": chat_response.get("sources", []),
                "confidence_score": chat_response.get("confidence_score")
            },
            safety={
                "emergency_detected": False,
                "disclaimer": disclaimer,
                "urgent_care_recommended": False
            },
            recommendations={
                "suggested_action": suggested_action,
                "follow_up_questions": []
            },
            processing_time_ms=processing_time,
            session_id=request.session_id or f"consultation-{user_id}",
            agent_id=chat_response.get("agent_id", request.preferred_agent)
        )
        
    except ValidationError as e:
        logger.error(f"❌ Validation error in medical consultation: {e.message}")
        raise HTTPException(status_code=422, detail=e.message)
        
    except ProcessingError as e:
        logger.error(f"❌ Processing error in medical consultation: {e.message}")
        raise HTTPException(status_code=500, detail=e.message)
        
    except Exception as e:
        logger.error(f"❌ Unexpected error in medical consultation: {str(e)}")
        logger.error(f"❌ Error traceback: {traceback.format_exc()}")
        raise HTTPException(
            status_code=500, 
            detail=f"Medical consultation failed: {str(e)}"
        )

File: backend/routes/medicalConsultation.js
import express from 'express';
import { verifyToken } from '../middleware/auth.js';
import { errorLogger } from '../agent_utils/shared/logger.js';
import { AgentService } from '../agent_utils/core/agentService.js';
import { TxAgentRAGService } from '../lib/services/TxAgentRAGService.js';
import axios from 'axios';

export function createMedicalConsultationRouter(supabaseClient) {
  const router = express.Router();
  router.use(verifyToken);
  
  const agentService = new AgentService(supabaseClient);

  // Emergency keywords for Phase 1
  const emergencyKeywords = [
    'chest pain', 'difficulty breathing', 'severe bleeding', 'unconscious',
    'heart attack', 'stroke', 'seizure', 'severe allergic reaction',
    'suicidal thoughts', 'overdose', 'can\'t breathe', 'choking',
    'severe headache', 'loss of consciousness', 'severe abdominal pain',
    'severe burns', 'poisoning', 'drug overdose', 'suicide', 'kill myself'
  ];

  const detectEmergency = (text) => {
    const lowerText = text.toLowerCase();
    const detectedKeywords = emergencyKeywords.filter(keyword =>
      lowerText.includes(keyword)
    );
    
    return {
      isEmergency: detectedKeywords.length > 0,
      confidence: detectedKeywords.length > 0 ? 'high' : 'low',
      detectedKeywords
    };
  };

  // ✅ NEW: OpenAI API call function (unchanged)
  const callOpenAI = async (query, context, userProfile) => {
    if (!process.env.OPENAI_API_KEY) {
      throw new Error('OpenAI API key not configured');
    }

    // Construct system prompt based on context
    let systemPrompt = `You are a helpful medical AI assistant. Provide accurate, evidence-based medical information while being clear that you are not a substitute for professional medical advice.

IMPORTANT DISCLAIMERS:
- Always remind users that this information is for educational purposes only
- Recommend consulting healthcare professionals for personalized medical advice
- Never provide emergency medical advice - direct users to emergency services for urgent situations

Guidelines:
- Be empathetic and understanding
- Use clear, accessible language
- Cite general medical knowledge when appropriate
- Acknowledge limitations of AI medical advice`;

    // Add user profile context if available
    if (userProfile) {
      systemPrompt += `\n\nUser Profile Context:`;
      if (userProfile.age) systemPrompt += `\n- Age: ${userProfile.age}`;
      if (userProfile.gender) systemPrompt += `\n- Gender: ${userProfile.gender}`;
      if (userProfile.conditions && userProfile.conditions.length > 0) {
        systemPrompt += `\n- Medical Conditions: ${userProfile.conditions.join(', ')}`;
      }
      if (userProfile.medications && userProfile.medications.length > 0) {
        systemPrompt += `\n- Current Medications: ${userProfile.medications.join(', ')}`;
      }
      if (userProfile.allergies && userProfile.allergies.length > 0) {
        systemPrompt += `\n- Known Allergies: ${userProfile.allergies.join(', ')}`;
      }
      systemPrompt += `\n\nPlease consider this profile information when providing advice, but always recommend consulting with their healthcare provider for personalized guidance.`;
    }

    // Prepare conversation history
    const messages = [
      { role: 'system', content: systemPrompt }
    ];

    // Add conversation history if available
    if (context?.conversation_history && Array.isArray(context.conversation_history)) {
      context.conversation_history.forEach(msg => {
        if (msg.type === 'user') {
          messages.push({ role: 'user', content: msg.content });
        } else if (msg.type === 'assistant') {
          messages.push({ role: 'assistant', content: msg.content });
        }
      });
    }

    // Add current query
    messages.push({ role: 'user', content: query });

    try {
      const response = await axios.post(
        'https://api.openai.com/v1/chat/completions',
        {
          model: 'gpt-4',
          messages: messages,
          max_tokens: 800,
          temperature: 0.7,
          presence_penalty: 0.1,
          frequency_penalty: 0.1
        },
        {
          headers: {
            'Authorization': `Bearer ${process.env.OPENAI_API_KEY}`,
            'Content-Type': 'application/json'
          },
          timeout: 30000
        }
      );

      return {
        text: response.data.choices[0].message.content,
        confidence: 0.8, // Default confidence for OpenAI responses
        usage: response.data.usage,
        model: 'gpt-4'
      };

    } catch (error) {
      errorLogger.error('OpenAI API call failed', error, {
        error_response: error.response?.data,
        error_status: error.response?.status,
        component: 'MedicalConsultation'
      });
      
      if (error.response?.status === 429) {
        throw new Error('OpenAI rate limit exceeded. Please try again later.');
      } else if (error.response?.status === 401) {
        throw new Error('OpenAI API authentication failed. Please check configuration.');
      } else {
        throw new Error(`OpenAI API error: ${error.message}`);
      }
    }
  };

  router.post('/medical-consultation', async (req, res) => {
    const startTime = Date.now();
    let userId = req.userId;

    try {
      const { 
        query, 
        context, 
        session_id, 
        preferred_agent = 'txagent', // ✅ NEW: Default to TxAgent
        include_voice = false,       // ✅ NEW: Voice generation flag
        include_video = false        // ✅ NEW: Video generation flag (future)
      } = req.body;

      if (!query || typeof query !== 'string') {
        return res.status(400).json({
          error: 'Query is required and must be a string',
          code: 'INVALID_QUERY'
        });
      }

      errorLogger.info('Medical consultation request started', {
        userId,
        queryLength: query.length,
        queryPreview: query.substring(0, 100),
        hasContext: !!context,
        sessionId: session_id,
        preferredAgent: preferred_agent, // ✅ NEW: Log preferred agent
        includeVoice: include_voice,
        includeVideo: include_video,
        component: 'MedicalConsultation'
      });

      // Phase 1: Emergency Detection (applies to both agents)
      const emergencyCheck = detectEmergency(query);
      
      if (emergencyCheck.isEmergency) {
        errorLogger.warn('Emergency detected in consultation', {
          userId,
          detectedKeywords: emergencyCheck.detectedKeywords,
          query: query.substring(0, 200),
          preferredAgent: preferred_agent,
          component: 'MedicalConsultation'
        });

        // Log emergency consultation
        await supabaseClient
          .from('medical_consultations')
          .insert({
            user_id: userId,
            session_id: session_id || 'emergency-session',
            query: query,
            response: 'Emergency detected - immediate medical attention recommended',
            emergency_detected: true,
            consultation_type: 'emergency_detection',
            processing_time: Date.now() - startTime,
            context_used: { 
              emergency_keywords: emergencyCheck.detectedKeywords,
              preferred_agent: preferred_agent
            }
          });

        return res.json({
          response: {
            text: 'I\'ve detected that you may be experiencing a medical emergency. Please contact emergency services immediately (call 911) or go to the nearest emergency room. This system cannot provide emergency medical care.',
            confidence_score: 0.95
          },
          safety: {
            emergency_detected: true,
            disclaimer: 'This is not a substitute for professional medical advice. In case of emergency, contact emergency services immediately.',
            urgent_care_recommended: true
          },
          recommendations: {
            suggested_action: 'Contact emergency services immediately (911)',
            follow_up_questions: []
          },
          processing_time_ms: Date.now() - startTime,
          session_id: session_id || 'emergency-session',
          agent_id: preferred_agent
        });
      }

      // ✅ NEW: Route based on preferred_agent
      let consultationResponse;
      let agentId;
      let processingDetails = {};

      if (preferred_agent === 'openai') {
        // ✅ NEW: OpenAI Route (no RAG - uses general knowledge)
        errorLogger.info('Routing consultation to OpenAI (no RAG)', {
          userId,
          queryLength: query.length,
          hasUserProfile: !!(context?.user_profile),
          component: 'MedicalConsultation'
        });

        try {
          const openAIResponse = await callOpenAI(
            query, 
            context, 
            context?.user_profile
          );

          consultationResponse = {
            text: openAIResponse.text,
            sources: [], // OpenAI doesn't use document sources
            confidence_score: openAIResponse.confidence
          };

          agentId = 'openai';
          processingDetails = {
            model: openAIResponse.model,
            usage: openAIResponse.usage,
            agent_type: 'openai',
            rag_used: false
          };

          errorLogger.info('OpenAI consultation completed successfully', {
            userId,
            responseLength: openAIResponse.text.length,
            tokensUsed: openAIResponse.usage?.total_tokens,
            model: openAIResponse.model,
            component: 'MedicalConsultation'
          });

        } catch (error) {
          errorLogger.error('OpenAI consultation failed', error, {
            userId,
            query: query.substring(0, 100),
            component: 'MedicalConsultation'
          });

          return res.status(500).json({
            error: 'OpenAI consultation failed. Please try again or switch to TxAgent.',
            code: 'OPENAI_CONSULTATION_FAILED',
            details: error.message,
            processing_time_ms: Date.now() - startTime
          });
        }

      } else {
        // ✅ ENHANCED: TxAgent Route with RAG Integration
        errorLogger.info('Routing consultation to TxAgent with RAG', {
          userId,
          queryLength: query.length,
          component: 'MedicalConsultation'
        });

        // Get active agent from database
        const agent = await agentService.getActiveAgent(userId);
        
        if (!agent || !agent.session_data?.runpod_endpoint) {
          errorLogger.warn('No active TxAgent found for consultation', {
            userId,
            hasAgent: !!agent,
            component: 'MedicalConsultation'
          });

          return res.status(503).json({
            error: 'TxAgent service is not available. Please ensure TxAgent is running or try using OpenAI.',
            code: 'TXAGENT_SERVICE_UNAVAILABLE',
            suggestions: [
              'Start TxAgent from the Monitor page',
              'Switch to OpenAI by setting preferred_agent to "openai"'
            ]
          });
        }

        // ✅ NEW: Initialize TxAgent RAG Service
        const txAgentUrl = agent.session_data.runpod_endpoint.replace(/\/+$/, '');
        const ragService = new TxAgentRAGService(
          supabaseClient,
          txAgentUrl,
          req.headers.authorization
        );

        try {
          // ✅ NEW: Perform RAG workflow
          errorLogger.info('Starting RAG workflow for TxAgent consultation', {
            userId,
            agentId: agent.id,
            txAgentUrl,
            queryLength: query.length,
            component: 'MedicalConsultation'
          });

          const ragResult = await ragService.performRAG(query, 5, 0.7);

          // ✅ NEW: Create augmented prompt with RAG context
          const augmentedQuery = ragService.createAugmentedPrompt(
            query,
            ragResult.context,
            context?.user_profile,
            context?.conversation_history
          );

          errorLogger.info('RAG workflow completed, sending augmented query to TxAgent', {
            userId,
            agentId: agent.id,
            originalQueryLength: query.length,
            augmentedQueryLength: augmentedQuery.length,
            documentsFound: ragResult.documentsFound,
            sourcesCount: ragResult.sources.length,
            component: 'MedicalConsultation'
          });

          // ✅ ENHANCED: Send augmented query to TxAgent
          const timeoutMs = parseInt(process.env.TXAGENT_TIMEOUT) || 120000;
          
          const txAgentResponse = await fetch(`${txAgentUrl}/chat`, {
            method: 'POST',
            headers: {
              'Authorization': req.headers.authorization,
              'Content-Type': 'application/json'
            },
            body: JSON.stringify({
              query: augmentedQuery, // ✅ NEW: Use augmented query with RAG context
              history: context?.conversation_history || [],
              top_k: 5,
              temperature: 0.7,
              stream: false,
              context: context // Pass full context including user_profile
            }),
            signal: AbortSignal.timeout(timeoutMs)
          });

          if (!txAgentResponse.ok) {
            const errorText = await txAgentResponse.text();
            throw new Error(`TxAgent responded with status ${txAgentResponse.status}: ${errorText.substring(0, 200)}`);
          }

          const txAgentData = await txAgentResponse.json();

          // ✅ NEW: Combine TxAgent response with RAG sources
          consultationResponse = {
            text: txAgentData.response,
            sources: ragResult.sources, // ✅ NEW: Use RAG sources instead of TxAgent sources
            confidence_score: txAgentData.confidence_score
          };

          agentId = 'txagent';
          processingDetails = {
            agent_id: agent.id,
            sources_count: ragResult.sources.length,
            documents_found: ragResult.documentsFound,
            processing_time: txAgentData.processing_time,
            model: txAgentData.model || 'BioBERT',
            agent_type: 'txagent',
            rag_used: true,
            rag_error: ragResult.error || null
          };

          errorLogger.info('TxAgent consultation with RAG completed successfully', {
            userId,
            processingTime: txAgentData.processing_time,
            sourcesCount: ragResult.sources.length,
            documentsFound: ragResult.documentsFound,
            agentId: agent.id,
            component: 'MedicalConsultation'
          });

        } catch (error) {
          // ✅ ENHANCED: Detailed error categorization
          let enhancedError = error;
          let errorCode = 'TXAGENT_CONSULTATION_FAILED';
          
          if (error.name === 'AbortError' || error.name === 'TimeoutError') {
            enhancedError = new Error(`TxAgent request timed out after ${timeoutMs}ms. The container may be overloaded or unresponsive.`);
            errorCode = 'TXAGENT_TIMEOUT';
          } else if (error.message.includes('fetch failed')) {
            enhancedError = new Error(`Network error connecting to TxAgent container at ${txAgentUrl}. The container may be unreachable.`);
            errorCode = 'TXAGENT_NETWORK_ERROR';
          } else if (error.message.includes('RAG')) {
            enhancedError = new Error(`RAG workflow failed: ${error.message}. Falling back to basic TxAgent response.`);
            errorCode = 'TXAGENT_RAG_FAILED';
          }
          
          errorLogger.error('TxAgent consultation with RAG failed', enhancedError, {
            userId,
            query: query.substring(0, 100),
            agentId: agent.id,
            txAgentUrl,
            errorName: error.name,
            errorMessage: error.message,
            errorStack: error.stack,
            timeoutMs,
            component: 'MedicalConsultation'
          });

          return res.status(502).json({
            error: enhancedError.message,
            code: errorCode,
            details: error.message,
            processing_time_ms: Date.now() - startTime,
            suggestions: [
              'Check TxAgent container status',
              'Try using OpenAI by setting preferred_agent to "openai"'
            ]
          });
        }
      }

      // ✅ UPDATED: Log consultation with enhanced RAG information
      const consultationRecord = {
        user_id: userId,
        session_id: session_id || agentId,
        query: query,
        response: consultationResponse.text,
        sources: consultationResponse.sources || [],
        consultation_type: `${agentId}_consultation`, // ✅ NEW: Include agent type
        processing_time: Date.now() - startTime,
        emergency_detected: false,
        context_used: {
          ...processingDetails,
          preferred_agent: preferred_agent,
          has_user_profile: !!(context?.user_profile),
          conversation_history_length: context?.conversation_history?.length || 0
        },
        confidence_score: consultationResponse.confidence_score || null,
        recommendations: {
          suggested_action: 'Consult with healthcare provider for personalized advice',
          follow_up_questions: []
        }
      };

      await supabaseClient
        .from('medical_consultations')
        .insert(consultationRecord);

      errorLogger.info('Medical consultation completed successfully', {
        userId,
        processingTime: Date.now() - startTime,
        sourcesCount: consultationResponse.sources?.length || 0,
        agentUsed: agentId,
        preferredAgent: preferred_agent,
        ragUsed: processingDetails.rag_used || false,
        component: 'MedicalConsultation'
      });

      // ✅ UPDATED: Return response with enhanced RAG information
      res.json({
        response: {
          text: consultationResponse.text,
          sources: consultationResponse.sources || [],
          confidence_score: consultationResponse.confidence_score
        },
        safety: {
          emergency_detected: false,
          disclaimer: 'This information is for educational purposes only and is not a substitute for professional medical advice, diagnosis, or treatment.',
          urgent_care_recommended: false
        },
        recommendations: consultationRecord.recommendations,
        processing_time_ms: Date.now() - startTime,
        session_id: session_id || agentId,
        agent_id: agentId, // ✅ NEW: Include which agent was actually used
        preferred_agent: preferred_agent, // ✅ NEW: Include what was requested
        rag_info: { // ✅ NEW: Include RAG information
          used: processingDetails.rag_used || false,
          documents_found: processingDetails.documents_found || 0,
          sources_count: processingDetails.sources_count || 0
        }
      });

    } catch (error) {
      const processingTime = Date.now() - startTime;
      
      errorLogger.error('Medical consultation failed', error, {
        userId,
        processingTime,
        query: req.body.query?.substring(0, 100),
        preferredAgent: req.body.preferred_agent,
        component: 'MedicalConsultation'
      });

      res.status(500).json({
        error: 'Medical consultation failed. Please try again.',
        code: 'CONSULTATION_FAILED',
        processing_time_ms: processingTime,
        preferred_agent: req.body.preferred_agent
      });
    }
  });

  return router;
}